Core/UsageSegmentationCalculator.swift

```diff
@@ -0,0 +1,372 @@
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+protocol UsageSegmentationCalculatorMaking {
+
+    /// Creates a calculator with the provided initial state
+    func make(installAtb: Atb) -> UsageSegmentationCalculating
+
+}
+
+/// The calculator does not store state, so to get the latest values call it for each previously seen ATB.
+protocol UsageSegmentationCalculating {
+
+    func processAtb(_ atb: Atb, forActivityType activityType: UsageActivityType) -> [String: String]?
+
+}
+
+final class UsageSegmentationCalculatorFactory: UsageSegmentationCalculatorMaking {
+
+    func make(installAtb: Atb) -> any UsageSegmentationCalculating {
+        return UsageSegmentationCalculator(installAtb: installAtb)
+    }
+
+}
+
+final class DefaultCalculatorFactory: UsageSegmentationCalculatorMaking {
+
+    func make(installAtb: Atb) -> any UsageSegmentationCalculating {
+        return UsageSegmentationCalculator(installAtb: installAtb)
+    }
+
+}
+
+/// This implementation is based on https://dub.duckduckgo.com/flawrence/felix-jupyter-modules/blob/master/segments_reference.py and has been written to try and 
+///   resemble the original code as closely as possible.
+///
+/// * Some general terminology changes:
+///   * new_set_atb => atb
+///   * atb_cohort => installAtb
+///
+/// * Commented code starting with # indicate comments copied over from the reference implementation.
+///
+/// * The code is arranged so that the public function which is the one that matters the most is the first thing you read.  Private functions are added to the end as they are encounted during the process of coverting the Python to Swift.
+///
+/// * It was agreed that for the purpose of comparison and calcuation that if previousAtb is nil we can use the installAtb
+///
+final class UsageSegmentationCalculator: UsageSegmentationCalculating {
+
+    enum Params {
+
+        static let activityType = "activity_type"
+        static let newSetAtb = "new_set_atb"
+        static let segmentsToday = "segments_today"
+        static let segmentsPreviousWeek = "segments_prev_week"
+        static let segmentsPreviousMonthPrefix = "segments_prev_month_"
+        static let countAsWAU = "count_as_wau"
+        static let countAsMAUn = "count_as_mau_n"
+
+    }
+
+    let installAtb: Atb
+    var usageHistory = [Atb]()
+    var previousAtb: Atb?
+    var previousWAUSegments = [String]()
+    var previousMAUSegments = Array(repeating: Array(repeating: "", count: 4), count: 4)
```

Comment:
Do all of these properties need to be internal? Could they be `private` or `private (set)`. (Checked `usageHistory` and it could but unsure of the rest).

Reply:
Yeah, they can and probably should be private.